File: lua/src/arrays.lua
-- Module for demonstrating arrays and tables in Lua
local M = {}

function M.demonstrate_arrays()
    print("=== Arrays and Tables ===")
    
    -- Simple array (1-based indexing in Lua)
    print("Simple array:")
    local numbers = {1, 2, 3, 4, 5}
    for i = 1, #numbers do
        print(string.format("numbers[%d] = %d", i, numbers[i]))
    end
    
    -- Array operations
    print("\nArray operations:")
    table.insert(numbers, 6)  -- Add to end
    print("After insert(6):", table.concat(numbers, ", "))
    
    table.remove(numbers)     -- Remove from end
    print("After remove:", table.concat(numbers, ", "))
    
    table.insert(numbers, 2, 10)  -- Insert at position 2
    print("After insert(2, 10):", table.concat(numbers, ", "))
    
    -- Table as dictionary
    print("\nTable as dictionary:")
    local person = {
        name = "Alice",
        age = 25,
        city = "Wonderland"
    }
    for key, value in pairs(person) do
        print(string.format("%s = %s", key, value))
    end
    
    -- Nested tables
    print("\nNested tables:")
    local matrix = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    }
    for i = 1, #matrix do
        print(string.format("Row %d: %s", i, table.concat(matrix[i], ", ")))
    end
    
    -- Mixed table
    print("\nMixed table:")
    local mixed = {
        10,
        20,
        name = "Bob",
        [true] = "boolean key",
        ["key"] = "string key"
    }
    print("Numeric indexes:")
    for i = 1, #mixed do
        print(string.format("[%d] = %s", i, mixed[i]))
    end
    print("Other keys:")
    print("name =", mixed.name)
    print("[true] =", mixed[true])
    print('["key"] =', mixed["key"])
    
    -- Table methods
    print("\nTable methods:")
    local fruits = {"banana", "apple", "orange", "grape"}
    print("Original:", table.concat(fruits, ", "))
    table.sort(fruits)
    print("Sorted:", table.concat(fruits, ", "))
    
    -- Creating a 2D array
    print("\n2D array with 3x3 zero matrix:")
    local matrix2d = {}
    for i = 1, 3 do
        matrix2d[i] = {}
        for j = 1, 3 do
            matrix2d[i][j] = 0
        end
    end
    for i = 1, 3 do
        print(table.concat(matrix2d[i], " "))
    end
    
    print() -- Empty line
end

return M

File: nim/src/arrays.nim
import std/[tables, sets]

proc demonstrateArrays*() =
  echo "=== Arrays and Collections ==="
  
  # Arrays (fixed size)
  echo "Arrays:"
  var arr: array[5, int] = [1, 2, 3, 4, 5]
  echo "Original array: ", arr
  arr[0] = 10
  echo "After modification: ", arr
  
  # Array operations
  echo "\nArray operations:"
  echo "First element: ", arr[0]
  echo "Last element: ", arr[^1]
  echo "Length: ", arr.len
  
  # Multi-dimensional arrays
  var matrix: array[3, array[3, int]]
  for i in 0..2:
    for j in 0..2:
      matrix[i][j] = i * 3 + j + 1
  
  echo "\nMatrix (2D array):"
  for row in matrix:
    echo row
  
  # Sequences (dynamic arrays)
  echo "\nSequences:"
  var seq1 = @[1, 2, 3, 4, 5]
  echo "Original sequence: ", seq1
  
  # Sequence operations
  seq1.add(6)
  echo "After adding 6: ", seq1
  seq1.delete(0)
  echo "After deleting first element: ", seq1
  
  # Sequence methods
  echo "First element: ", seq1[0]
  echo "Length: ", seq1.len
  echo "Capacity: ", seq1.cap
  
  # Slicing
  echo "\nSlicing:"
  echo "Slice [1..3]: ", seq1[1..3]
  echo "Slice [0..^2]: ", seq1[0..^2]  # All but last element
  
  # Tables (hash maps)
  echo "\nTables:"
  var scores = {
    "Alice": 95,
    "Bob": 87,
    "Charlie": 92
  }.toTable
  
  echo "Original scores: ", scores
  scores["David"] = 88
  echo "After adding David: ", scores
  echo "Alice's score: ", scores["Alice"]
  
  # Table operations
  echo "Has Alice? ", scores.hasKey("Alice")
  echo "Keys: ", scores.keys.toSeq
  echo "Values: ", scores.values.toSeq
  
  # OrderedTable (maintains insertion order)
  echo "\nOrderedTable:"
  var orderedScores = initOrderedTable[string, int]()
  orderedScores["First"] = 1
  orderedScores["Second"] = 2
  orderedScores["Third"] = 3
  echo orderedScores
  
  # Sets
  echo "\nSets:"
  var numberSet = [1, 2, 3, 2, 1].toHashSet
  echo "Original set: ", numberSet
  numberSet.incl(4)
  echo "After adding 4: ", numberSet
  numberSet.excl(1)
  echo "After removing 1: ", numberSet
  
  # Set operations
  var
    set1 = [1, 2, 3].toHashSet
    set2 = [2, 3, 4].toHashSet
  
  echo "\nSet operations:"
  echo "Set1: ", set1
  echo "Set2: ", set2
  echo "Union: ", set1 + set2
  echo "Intersection: ", set1 * set2
  echo "Difference: ", set1 - set2
  
  # Tuples
  echo "\nTuples:"
  type Person = tuple[name: string, age: int]
  let
    person1: Person = (name: "Alice", age: 25)
    person2 = ("Bob", 30)  # Anonymous tuple
  
  echo "Named tuple: ", person1
  echo "Anonymous tuple: ", person2
  echo "Name: ", person1.name
  echo "Age: ", person1.age
  
  # Tuple unpacking
  let (name, age) = person1
  echo "Unpacked - Name: ", name, ", Age: ", age
  
  echo "" # Empty line